components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,129 @@ import "google/api/annotations.proto";
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // SetActiveClient to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Typically called by the IDE.
+  // Will response what client should do like `open` `preview`
+  rpc SetActiveClient(SetActiveClientRequest)
+      returns (stream SetActiveClientResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/active-client"
+    };
+  }
+
+  rpc Action(ActionRequest) returns (ActionResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/action"
+    };
+  }
+
+  rpc ActionRespond(ActionRespondRequest) returns (ActionRespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/action-respond"
+    };
+  }
 }
 
 message NotifyRequest {
-    enum Level {
-        ERROR = 0;
-        WARNING = 1;
-        INFO = 2;
-    }
-    Level level = 1;
-    string message = 2;
-    // if actions are empty, Notify will return immediately
-    repeated string actions = 3;
+  enum Level {
+    ERROR = 0;
+    WARNING = 1;
+    INFO = 2;
+  }
+  Level level = 1;
+  string message = 2;
+  // if actions are empty, Notify will return immediately
+  repeated string actions = 3;
 }
 
 message NotifyResponse {
-    // action chosen by the user or empty string if cancelled
-    string action = 1;
+  // action chosen by the user or empty string if cancelled
+  string action = 1;
 }
 
 message SubscribeRequest {}
 
 message SubscribeResponse {
-    uint64 requestId = 1;
-    NotifyRequest request = 2;
+  uint64 requestId = 1;
+  NotifyRequest request = 2;
 }
 
 message RespondRequest {
-    uint64 requestId = 1;
-    NotifyResponse response = 2;
+  uint64 requestId = 1;
+  NotifyResponse response = 2;
+}
+
+message RespondResponse {}
+
+enum ActionMethod {
+  // open a file in editor
+  OPEN = 0;
+  // ask editor to open a URL in its preview
+  PREVIEW = 1;
+}
+
+message ActionRequest {
+  ActionMethod method = 1;
+
+  message OpenData {
+    repeated string urls = 1;
+    // wait until all opened files are closed
+    bool await = 2;
+  }
+  OpenData open = 2;
+
+  message PreviewData {
+    string url = 1;
+    // open the URL in a new browser tab
+    bool external = 2;
+  }
+  PreviewData preview = 3;
+}
+
+message ActionResponse {
+  uint64 code = 1;
```

Comment:
Why do we need `code` and `message`?

Reply:
Consider it like http response, cmd like `code open` something's failed in some reason, we use code to check if it's ok, use message to notice users why it's not ok etc

Reply:
I think you are still affected by thiking about exec. Clients won't exec anthing, they will call internal APIs to do something. There is no code.

Reply:
No exec, code is something like status code